### DISCUSSION TITLE
Multiplatform

### DISCUSSION THREAD

**Discussion by svenjacobs** [svenjacobs/reveal#40] [Ideas]
Everyone is talking about [Compose Multiplatform](https://www.jetbrains.com/lp/compose-multiplatform/) right now.
This is a discussion to figure out how Reveal can be ported to multiplatform.

<hr/>

The biggest challenge I currently see is the dependence on Android's `ComposeView` and `Context` in the [Fullscreen](https://github.com/svenjacobs/reveal/blob/main/reveal-core/src/main/kotlin/com/svenjacobs/reveal/internal/fullscreen/Fullscreen.kt) composable, which renders the Reveal effect on top of all other composables, regardless where in the UI hierarchy it's placed.

**Comment by svenjacobs** [comment]
> The biggest challenge I currently see is the dependence on Android's ComposeView and Context in the [Fullscreen](https://github.com/svenjacobs/reveal/blob/main/reveal-core/src/main/kotlin/com/svenjacobs/reveal/internal/fullscreen/Fullscreen.kt) composable, which renders the Reveal effect on top of all other composables, regardless where in the UI hierarchy it's placed.

This was introduced with PR #25 in version [0.0.14](https://github.com/svenjacobs/reveal/releases/tag/v0.0.14). Before this change the `Reveal` composable needed to be one of the top most composables of the whole Compose hierarchy in order to render the "full screen" effect over all other composables. This also meant that `RevealScope`, which is required for the `revealable()` modifier, needed to be passed down to the composable where needed. The change simplified the usage of Reveal because `Reveal` could be just added where needed, in a separate module for example, but the effect would still render full screen.

However this has introduced a dependency to Android's `Context`. In order to remove this dependency again and pave the way for multiplatform support, we would have to change the behaviour of Reveal to how it was before this change. This is obviously a **breaking change** and would call for a new major version (`2.0`).

In summary, this would mean:

- `Reveal` needs to be declared as one of the top most composables. Reveal becomes a dependency of the whole project.
- `RevealScope` needs to be passed down to the place where the `revealable()` modifier is applied. This will inevitable introduce some noise and boiler plate code. Since scopes are often used in Compose UI (like `BoxScope`, `RowScope`, etc.), this will likely introduce "scope conflicts" where one of the scopes need to be passed as an explicit parameter instead of a receiver. Since [multiple context receivers](https://github.com/Kotlin/KEEP/blob/master/proposals/context-receivers.md) are still experimental and will likely become stable after Kotlin 2.0 — probably in 2024 _(just guessing)_ — we have no nice solution for this until Kotlin supports it.

@DATL4G what do you think?

**Reply by DatL4g** [reply]
I think the `Reveal` as one of the top composables is fine.
However I have another plan for the `RevealScope`, we just get rid of the scoping :)
This way we don't face any conflicts, if other Composables are called

When the `Reveal` method is called, we internally create a static composition local, when the `revealable` modifier is applied we require the scope as parameter.
To get the parameter the user can call like a `rememberRevealScope()` method, maybe we can call it as default "value" in `revealable` already.

Proof-of-concept:

```kotlin
fun main() {
	singleWindowApplication(
		title = "Example"
	) {
		Theme {
			Reveal {
				content()
			}
		}
	}
}

@Composable
fun content() {
	Scaffold {
		// further nesting etc.
		val revealScope = rememberRevealScope()

		Text(
			text = "Example",
			modifier = Modifier.revealable(scope = revealScope, key = "key")
		)
	}
}
```

**Reply by svenjacobs** [reply]
```kotlin
val revealScope = rememberRevealScope()

Text(
    text = "Example",
    modifier = Modifier.revealable(scope = revealScope, key = "key")
)
```

Currently the `revealable()` modifier is declared as an extension function in the scope of [RevealScope](https://github.com/svenjacobs/reveal/blob/459025977262d6f66c263d1f47cb77e54dd81a3a/reveal-core/src/main/kotlin/com/svenjacobs/reveal/RevealScope.kt). The beauty of this is that the modifier is only available inside the scope. Having to remember a scope is quite unusual. I haven't seen that in Compose UI before. I designed Reveal like Compose UI is designed so that it feels familiar. I think we should not change these core principals or users might be confused by the different handling of Reveal.

Internally the `RevealScope` only provides the current instance of `RevealState` to the modifier. What we could do is provide an alternative "overload" of the `revealable()` modifier that is not bound to `RevealScope` but requires a `RevealState` argument. That way the users can decide whether they pass down the scope **or** the state.

**Reply by DatL4g** [reply]
I've seen calling and passing the scope like this already, but if it's just for providing the state then it's much better to do it the way you meant yes.

**Reply by svenjacobs** [reply]
As an intermediate step towards Compose Multiplatform compatibility, in [this PR](https://github.com/svenjacobs/reveal/pull/60) I removed the `Context` dependency from the core artifact and moved it to a new `reveal-compat-android` artifact. I introduced a `RevealOverlayInserter`, which declares a strategy of how the overlay is added to the composition. The current backwards compatible strategy is to still use the `FullscreenRevealOverlayInserter`, which requires an Android `Context`. Also I added the unbounded modifier variants that take a `state` argument as mentioned above.

All modules are still Android libraries and use Jetpack Compose, however now it should be much easier to migrate all modules except `reveal-compat-android` to Compose Multiplatform.

If one would prepare their application for a future version of Reveal with Compose Multiplatform support now, they should just use a single instance of the `Reveal` composable at the root of the tree and specify `InPlaceRevealOverlayInserter` for the `overlayInserter` argument.

**Reply by DatL4g** [reply]
Alright hit me up if you need help towards the multiplatform migration

**Reply by svenjacobs** [reply]
Thinking about this a bit more, I noticed a few problems with only having one `Reveal` composable in an application:

Currently you define the click listener on a revealable item or the overlay via `onRevealableClick` and `onOverlayClick` of `Reveal`. Also you define the overlay content via `overlayContent` there as well. This is fine when you have multiple `Reveal` instances, one per "screen" of your application, probably spread over separate modules.

Now when you have only one `Reveal` in your whole application, you have to deal with **all** click events and overlay content of our whole application in a **single location**! Also you have to delegate click events somehow down to the screen where you actually want to handle the click, probably performing some business logic in a ViewModel or so. This is a violation of separation of concerns.

Therefor I think we should separate the composable that draws the overlay and the composable that handles the registration and interaction of revealable items for a single screen. I created a [RevealCanvas](https://github.com/svenjacobs/reveal/pull/63/files?w=1) as a proof of concept. There should only be a single `RevealCanvas` at the top of the composable tree in your application but there can be many `Reveal` instances. What do you think about this design @DATL4G?

**Reply by DatL4g** [reply]
LGTM makes way more sense

**Reply by svenjacobs** [reply]
Alright. I think this change will be a `2.0` because the behavior has changed significantly and the Multiplatform support will be a `3.0` later.

**Comment by svenjacobs** [comment]
I'm happy to announce that version [3.0](https://github.com/svenjacobs/reveal/releases/tag/v3.0.0) supports Compose Multiplatform!
